File: Backend/src/main/java/com/github/francisfire/anavis/controller/PrenotationController.java
package com.github.francisfire.anavis.controller;

import java.util.Set;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import com.github.francisfire.anavis.component.AccessCheckerComponent;
import com.github.francisfire.anavis.models.ActivePrenotation;
import com.github.francisfire.anavis.services.DonationReportServices;
import com.github.francisfire.anavis.services.PrenotationServices;

@RestController
@RequestMapping("api/prenotation")
public class PrenotationController {

	@Autowired
	private PrenotationServices prenotationServices;

	@Autowired
	private DonationReportServices donationReportServices;

	@SuppressWarnings("unused")
	@Autowired
	private AccessCheckerComponent accessCheckerComponent;

	@PreAuthorize("hasAnyAuthority('OFFICE')")
	@PostMapping("")
	public boolean createPrenotation(@RequestBody ActivePrenotation prenotation) {
		prenotation.setId(prenotation.getDonorMail() + "-" + prenotation.getOfficeMail() + "-"
				+ prenotation.getHour().toString());
		return prenotationServices.addPrenotation(prenotation);
	}

	@PreAuthorize("@accessCheckerComponent.sameUserId(principal, #officeMail)")
	@GetMapping("/office/{officeMail}")
	public Set<ActivePrenotation> getPrenotationsByOffice(@PathVariable("officeMail") String officeMail) {
		return prenotationServices.getPrenotationsByOffice(officeMail);
	}

	@PreAuthorize("@accessCheckerComponent.sameUserId(principal, #donorMail)")
	@GetMapping("/donor/{donorMail}")
	public Set<ActivePrenotation> getPrenotationsByDonor(@PathVariable("donorMail") String donorMail) {
		return prenotationServices.getPrenotationsByDonor(donorMail);
	}

	@PreAuthorize("@accessCheckerComponent.isPrenotationInstanceOwnedByOfficeId(principal, #prenotation)")
	@PutMapping("")
	public boolean updatePrenotation(@RequestBody ActivePrenotation prenotation) {
		return prenotationServices.updatePrenotation(prenotation);
	}

	@PreAuthorize("@accessCheckerComponent.isPrenotationOwnedByDonorId(principal, #prenotationId) or @accessCheckerComponent.isPrenotationOwnedByOfficeId(principal, #prenotationId)")
	@DeleteMapping("/{prenotationId}")
	public boolean removePrenotation(@PathVariable("prenotationId") String prenotationId) {
		return prenotationServices.removePrenotation(prenotationId);
	}

	@PreAuthorize("@accessCheckerComponent.isPrenotationOwnedByDonorId(principal, #prenotationId)")
	@PutMapping("/{prenotationId}/acceptChange")
	public boolean acceptPrenotationChange(@PathVariable("prenotationId") String prenotationId) {
		return prenotationServices.acceptPrenotationChange(prenotationId);
	}

	@PreAuthorize("@accessCheckerComponent.isPrenotationOwnedByDonorId(principal, #prenotationId)")
	@PutMapping("/{prenotationId}/denyChange")
	public boolean denyPrenotationChange(@PathVariable("prenotationId") String prenotationId) {
		return prenotationServices.denyPrenotationChange(prenotationId);
	}

	@PreAuthorize("@accessCheckerComponent.isPrenotationOwnedByOfficeId(principal, #prenotationId)")
	@PutMapping("/{prenotationId}/close")
	public boolean closePrenotation(@PathVariable("prenotationId") String prenotationId,
			@RequestParam("file") MultipartFile reportFile) {
		String reportId = donationReportServices.addReport(prenotationId, reportFile);
		return prenotationServices.closePrenotation(prenotationId, reportId);
	}
}

File: Frontend/lib/apicontrollers/prenotation_controller.dart
import 'dart:async';
import 'dart:convert';
import 'package:dio/dio.dart' as dio;
import 'package:anavis/models/activeprenotation.dart';
import 'package:anavis/providers/app_state.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

class PrenotationController {
  String _ip;
  String _baseUrl;
  PrenotationController(BuildContext context) {
    _ip = AppState().getIp();
    _baseUrl = "http://$_ip:8080/api/prenotation";
  }

  Future<String> createPrenotation(ActivePrenotation prenotation) async {
    http.Response res = await http.post(
      Uri.encodeFull("$_baseUrl"),
      body: json.encode({
        "id": prenotation.getId(),
        "officeMail": prenotation.getOfficeMail(),
        "donorMail": prenotation.getDonorMail(),
        "hour": prenotation.getHour(),
        "confirmed": prenotation.isConfirmed(),
      }),
      headers: {
        "content-type": "application/json",
        "accept": "application/json",
      },
    );
    return res.body;

    /*return await http.post(
      Uri.encodeFull("http://${_ipReference}:8080/api/prenotation"),
      body: json.encode({
        "id": prenotation.getId(),
        "officeMail": prenotation.getOfficeMail(),
        "donorMail": prenotation.getDonorMail(),
        "hour": prenotation.getHour(),
        "confirmed": prenotation.isConfirmed(),
      }),
      headers: {
        "content-type": "application/json",
        "accept": "application/json",
      },
    ).then((res) {
      _statusBody = res.body == 'true';
      notifyListeners();
    }).catchError((err) {
      _statusBody = false;
      notifyListeners();
    });*/
  }

  Future<String> getPrenotationsByOffice(String officeMail) async {
    http.Response res = await http.get("$_baseUrl/office/$officeMail");
    return res.body;
  }

  Future<String> getPrenotationsByDonor(String donorMail) async {
    http.Response res = await http.get("$_baseUrl/donor/$donorMail");
    return res.body;
  }

  Future<String> updatePrenotation(ActivePrenotation prenotation) async {
    http.Response res = await http.put(
      Uri.encodeFull("$_baseUrl"),
      body: json.encode({
        "id": prenotation.getId(),
        "officeMail": prenotation.getOfficeMail(),
        "donorMail": prenotation.getDonorMail(),
        "hour": prenotation.getHour(),
        "confirmed": prenotation.isConfirmed(),
      }),
      headers: {
        "content-type": "application/json",
        "accept": "application/json",
      },
    );
    return res.body;

    /*return await http.put(
      Uri.encodeFull("http://${_ipReference}:8080/api/prenotation"),
      body: json.encode({
        "id": prenotation.getId(),
        "officeMail": prenotation.getOfficeMail(),
        "donorMail": prenotation.getDonorMail(),
        "hour": prenotation.getHour(),
        "confirmed": prenotation.isConfirmed(),
      }),
      headers: {
        "content-type": "application/json",
        "accept": "application/json",
      },
    ).then((res) {
      _statusBody = res.body == 'true';
      notifyListeners();
    }).catchError((err) {
      _statusBody = false;
      notifyListeners();
    });*/
  }

  Future<String> removePrenotation(String prenotationId) async {
    http.Response res = await http.delete("$_baseUrl/$prenotationId");
    return res.body;
  }

  Future<String> acceptPrenotationChange(String prenotationId) async {
    http.Response res = await http.put("$_baseUrl/$prenotationId/acceptChange");
    return res.body;
  }

  Future<String> denyPrenotationChange(String prenotationId) async {
    http.Response res = await http.put("$_baseUrl/$prenotationId/denyChange");
    return res.body;
  }

  Future<String> closePrenotation(
      String prenotationId, dio.FormData formData) async {
    dio.Response<dynamic> res =
        await dio.Dio().put("$_baseUrl/$prenotationId/close", data: formData);
    return res.data.toString();
  }
}
